File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void BSTinorder(TreeNode* root, vector<int>& inorder){
        if(root == nullptr){
            return;
        }
        BSTinorder(root->left,inorder);
        inorder.push_back(root->val);
        BSTinorder(root->right,inorder);
    }
    int kthSmallest(TreeNode* root, int k) {
        vector<int> inorder;
        BSTinorder(root,inorder);
        return inorder[k-1];
    }
};

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        int count = 0;//counter for the number of nodes visited
        TreeNode curr = root;//start from the root
        while(curr != null){
            if(curr.left == null){
                //visit the current node
                count++;
                if(count == k){
                    return curr.val;//if it's the kth smallest, return its value
                }
                curr = curr.right;//move to the right child
            } 
            else{
                //find the rightmost node in the left subtree
                TreeNode p = curr.left;
                while(p.right != null && p.right != curr){
                    p = p.right;
                }
                //make the current node the right child of its predecessor
                if(p.right == null){
                    p.right = curr;//create a temporary link
                    curr = curr.left;//move to the left child
                } 
                else{
                    //revert the changes made to restore the original tree
                    p.right = null;//remove the temporary link
                    count++;//visit the current node
                    if(count == k){
                        return curr.val;//if it's the kth smallest, return its value
                    }
                    curr = curr.right;//move to the right child
                }
            }
        }

        return -1;
    }
}
